src/Oscoin/Data/OscoinTx.hs

```diff
@@ -0,0 +1,333 @@
+    -- ^ Unregister a project.
+    | TxAuthorize         (Address c) (Address c)
+    -- ^ Authorize an account to operate on a project.
+    | TxDeauthorize       (Address c) (Address c)
+    -- ^ De-authorize an account to operate on a project.
+    | TxCheckpoint        (Address c) (Crypto.Hash c) [Contribution c] [DependencyUpdate c]
+    -- ^ Checkpoint a project's state.
+    | TxUpdateContract    (Address c)
+    -- ^ Update a project's contract.
+    | TxTransfer          (Address c) (Address c) Balance
+    -- ^ Transfer balance between two accounts.
+    deriving (Generic)
+
+-- | Transaction output, included in the receipts tree. (Placeholder)
+data TxOutput = TxOutput
+
+-- | A transaction evaluation error.
+data TxError c =
+      ErrInsufficientBalance         Balance
+    -- ^ Insufficient balance in the account to execute transfer.
+    | ErrNotAuthorized               (Address c)
+    -- ^ This account is not authorized to execute the operation.
+    | ErrInvalidTransfer             Balance
+    -- ^ A balance transfer of this amount is not valid.
+    | ErrKeyNotFound                 StateKey
+    -- ^ The key was not found in the state.
+    | ErrTypeMismatch                StateKey
+    -- ^ The value under this key does not have the expected type.
+    | ErrProjectExists               (Address c)
+    -- ^ The project already exists in the state.
+    | ErrHandlerFailed               Error
+    -- ^ A contract handler failed and returned an error.
+    | ErrInvalidNonce                Nonce
+    -- ^ The supplied nonce does not match the account nonce.
+    | ErrInvalidFee                  Balance
+    -- ^ The transaction fee is not valid.
+    | ErrInvalidTx                   Text
+    -- ^ The transaction is not valid.
+    deriving (Generic, Eq, Ord)
+
+deriving instance (Show (PublicKey c), Show (Crypto.Hash c)) => Show (TxError c)
+
+-------------------------------------------------------------------------------
+
+-- | Make a transaction with the given payload.
+mkTx :: forall version flags c. (KnownNat version, KnownNat flags)
+     => TxPayload c -> Tx' version flags c
+mkTx p = Tx'
+    { txVersion = fromIntegral (natVal (Proxy @version))
+    , txFlags   = fromIntegral (natVal (Proxy @flags))
+    , txPayload = p
+    , txNetwork = Devnet
+    , txNonce   = 0
+    , txFee     = 0
+    , txBurn    = 0
+    }
+
+-- | Apply a transaction to the world state, and return either an error,
+-- or a new world state with the output of the transaction.
+applyTx
+    :: ( ByteArrayAccess (Crypto.ShortHash c)
+       , CBOR.Serialise (Crypto.ShortHash c)
+       , Ord (Signature c)
+       , Ord (PublicKey c)
+       , Ord (Crypto.Hash c)
+       )
+    => Tx c
+    -- ^ The transaction to apply.
+    -> Address c
+    -- ^ The author or \"signer\" of the transaction.
+    -> WorldState c
+    -- ^ The input world state.
+    -> Either (TxError c) (WorldState c, TxOutput)
+applyTx tx@Tx'{..} author ws = do
```

Comment:
I think we should at least have tests for this function.

Reply:
To do this properly will be a lot of code, which I'd rather handle in one of the subsequent PRs, since this code is not currently in use. We'll need generators for all payloads for both valid and invalid transactions, etc.